canopy-cover-enhancement/terra_ganEnhancement.py

```diff
@@ -0,0 +1,262 @@
+    return gray
+
+
+def MAC(im1, im2, im):  # main function: Multiscale Autocorrelation (MAC)
+    h, v, c = im1.shape
+    if c > 1:
+        im = np.matrix.round(rgb2gray(im))
+        im1 = np.matrix.round(rgb2gray(im1))
+        im2 = np.matrix.round(rgb2gray(im2))
+    # multiscale parameters
+    scales = np.array([2, 3, 5])
+    FM = np.zeros(len(scales))
+    for s in range(len(scales)):
+        im1[0: h - 1, :] = im[1:h, :]
+        im2[0: h - scales[s], :] = im[scales[s]:h, :]
+        dif = im * (im1 - im2)
+        FM[s] = np.mean(dif)
+    NRMAC = np.mean(FM)
+    return NRMAC
+
+
+# check how many percent of pix close to 255 or 0
+def check_saturation(img):
+    grayImg = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+
+    m1 = grayImg > SATURATE_THRESHOLD
+    m2 = grayImg < 20  # 20 is a threshold to classify low pixel value
+
+    over_rate = float(np.sum(m1)) / float(grayImg.size)
+    low_rate = float(np.sum(m2)) / float(grayImg.size)
+
+    return over_rate, low_rate
+
+
+# gen average pixel value from grayscale image
+def check_brightness(img):
+    grayImg = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+
+    aveValue = np.average(grayImg)
+
+    return aveValue
+
+
+# abandon low quality images, mask enhanced
+def gen_cc_enhanced(input_path, kernelSize=3):
+    img = cv2.imread(input_path)
+
+    # calculate image scores
+    over_rate, low_rate = check_saturation(img)
+
+    aveValue = check_brightness(img)
+
+    quality_score = getImageQuality(input_path)
+
+    # if low score, return None
+    # low_rate is percentage of low value pixels(lower than 20) in the grayscale image, if low_rate > 0.1, return
+    # aveValue is average pixel value of grayscale image, if aveValue lower than 30 or higher than 195, return
+    # quality_score is a score from Multiscale Autocorrelation (MAC), if quality_score lower than 13, return
+    if low_rate > 0.1 or aveValue < 30 or aveValue > 195 or quality_score < 13:
+        return None, None, None
+
+    # saturated image process
+    # over_rate is percentage of high value pixels(higher than SATUTATE_THRESHOLD) in the grayscale image, if over_rate > 0.15, try to fix it use gen_saturated_mask()
+    if over_rate > 0.15:
+        binMask = gen_saturated_mask(img, kernelSize)
+    else:  # nomal image process
+        binMask = gen_mask(img, kernelSize)
+
+    c = np.count_nonzero(binMask)
+    ratio = c / float(binMask.size)
+
+    rgbMask = gen_rgb_mask(img, binMask)
+
+    return ratio, binMask, rgbMask
+
+
+class ganEnhancementExtractor(TerrarefExtractor):
+
+    # TODO should we ignore winter wheat seasons?
+
+    def __init__(self):
+        super(ganEnhancementExtractor, self).__init__()
+
+        add_local_arguments(self.parser)
+
+        # assign local arguments
+        self.SATURATE_THRESHOLD = self.args.saturate_threshold
+        self.MAX_PIXEL_VAL = self.args.max_pixel_val
+        self.SMALL_AREA_THRESHOLD = self.args.small_area_threshold
```

Comment:
we can just lose these 3 and the parser command line arguments corresponding to them and just use the declared values above. i think you're right that if those values change significantly for other plants that would probably be considered a different extractor anyway.